### ISSUE TITLE
Xen fails booting using i7-9750H iGPU

### ISSUE THREAD

**Issue by JarrahG** [QubesOS/qubes-issues#5074]
### Qubes OS version:

R4.0.1, ISO and fully patched.

### Affected component(s) or functionality:
Xen boot. System boot does not get past the initial Xen loader.

---

### Steps to reproduce the behavior:
Start Qubes (either ISO or installed) on a laptop with an i7-9750H in EFI mode. On my system, leaving the iGPU enabled results in this behaviour. Disabling the iGPU and enabling only the Dedicated Nvidia card allows booting.

### Expected or desired behavior:
Xen is able to boot, allowing dom0/QubesOS to boot. 

### Actual behavior:
System never leaves the Xen loader. System either hangs at the loader with the following error message, or reboots after a number of minutes.

```
Xen 4.8.5-6.fc25 (c/s ) EFI loader
Using configuration file 'xen.cfg'
vmlinuz-4.14.119-2.pvops.qubes.x86_64: 0x000000002ecb5000-0x000000002f298c80
initramfs-4.14.119-2.pvops.qubes.x86_64.img: 0x000000002d6e1000-0x000000002ecb4cb2
0x0000:0x00:0x02.0x0: ROM: 0x10000 bytes at 0x3a4e9018
```

### General notes:

When set to the dedicated Nvidia GPU, the system is able to boot. However, as the card is not supported by Nouveau, this is not a useable workaround. This does however show that the GPU change is the deciding factor.

I have tested with the following system states:
* ISO/Fresh install: The exact system found the in 4.0.1 ISO.
* Current fully patched QubesOS.
* Kernel-latest fully patched QubesOS.

---

### I have consulted the following relevant [documentation](https://www.qubes-os.org/doc/):
https://github.com/QubesOS/qubes-issues/issues/794
https://www.qubes-os.org/doc/uefi-troubleshooting/
https://www.reddit.com/r/Qubes/comments/6f2kuu/tutorial_for_those_having_installationboot/
https://www.qubes-os.org/doc/intel-igfx-troubleshooting/
https://www.qubes-os.org/doc/nvidia-troubleshooting/

### I am aware of the following related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues:

https://github.com/QubesOS/qubes-issues/issues/4474 is related. However, two significant differences exist:
* This issue does not affect other distributions. Archlinux is able to boot without issue when the iGPU is enabled. 
* I was unable to get the system past the xen loader described above in iGPU mode. This issue seems to be having problems with X11.

**Comment by JarrahG** [comment]
A few updates and typos:
The CPU is the i7-9750H. Evidently I repeatedly mistyped the 7 and 9. 

UEFI update was provided by the vendor. However, this did not change the result, only the specific error values:
```
Xen 4.8.5-6.fc25 (c/s ) EFI loader
Using configuration file 'xen.cfg'
vmlinuz-4.14.119-2.pvops.qubes.x86_64: 0x00000000334d5000-0x0000000033AB4B20
initramfs-4.14.119-2.pvops.qubes.x86_64.img: 0x000000003217F1000-0x00000000334D4168
0x0000:0x00:0x02.0x0: ROM: 0x10000 bytes at 0x3e512018
```
It seems likely to me that this is a XEN bug.

**Comment by marmarek** [comment]
This may be indeed related to Xen bug, fixed [here](https://github.com/QubesOS/qubes-issues/issues/5074). Technically Xen didn't handled video framebuffer address above 4GB.
Unfortunately, reliably confirming this require source code change (more debug logging).
I'll let you know when a test build with those patches will be available.

**Comment by JarrahG** [comment]
Thanks for looking into this Marmarek, 

The link above links back to this thread. Is there a commit or PR that I can have a look at. 

I have a Qubes builder machine set up. If I can get the relevant commit to it, I should be able to build an ISO with the debug logging enabled to check whether the commit is relevant. 

Could you please direct me to the xen fix, as well as the location of the debug flag in the xen build?

**Comment by marmarek** [comment]
> The link above links back to this thread.

Oh, I've failed to paste the right link. Fixed now.

**Comment by JarrahG** [comment]
Thanks for the updated link marmarek. I've had to make a few minor changes, but vmm-xen is compiling with the attached patch. 
Only change is adding  " || defined(__XEN__)" as you recommended and changing the "__XEN_LATEST_INTERFACE_VERSION__" expected by the patch to match the current version found in qubes vmm-xen.

Interestingly, it compiles fine, outputting the relevant RPMs, but the build exits as a failure. 

```
Processing files: xen-qubes-vm-4.8.5-6.fc25.x86_64
Provides: config(xen-qubes-vm) = 2001:4.8.5-6.fc25 python2.7dist(xen) = 3.0 python2dist(xen) = 3.0 xen-qubes-vm = 2001:4.8.5-6.fc25 xen-qubes-vm(x86-64) = 2001:4.8.5-6.fc25 xen-qubes-vm-essentials = 2001:4.8.5-6.fc25
Requires(interp): /bin/sh /bin/sh /bin/sh
Requires(rpmlib): rpmlib(CompressedFileNames) <= 3.0.4-1 rpmlib(FileDigests) <= 4.6.0-1 rpmlib(PartialHardlinkSets) <= 4.0.4-1 rpmlib(PayloadFilesHavePrefix) <= 4.0-1
Requires(post): /bin/sh
Requires(preun): /bin/sh
Requires(postun): /bin/sh
Requires: /bin/bash /bin/sh libc.so.6()(64bit) libc.so.6(GLIBC_2.14)(64bit) libc.so.6(GLIBC_2.2.5)(64bit) libc.so.6(GLIBC_2.3)(64bit) libc.so.6(GLIBC_2.3.4)(64bit) libc.so.6(GLIBC_2.4)(64bit) libdl.so.2()(64bit) libpthread.so.0()(64bit) libpthread.so.0(GLIBC_2.2.5)(64bit) libpython2.7.so.1.0()(64bit) libsystemd.so.0()(64bit) libxenctrl.so.4.8()(64bit) libxenguest.so.4.8()(64bit) libxenlight.so.4.8.1()(64bit) libxenstore.so.3.0()(64bit) libxentoollog.so.1()(64bit) libxentoollog.so.1(VERS_1.0)(64bit) libxlutil.so.4.8()(64bit) libyajl.so.2()(64bit) python(abi) = 2.7 rtld(GNU_HASH)
Conflicts: xen
Processing files: xen-debuginfo-4.8.5-6.fc25.x86_64
Provides: xen-debuginfo = 2001:4.8.5-6.fc25 xen-debuginfo(x86-64) = 2001:4.8.5-6.fc25
Requires(rpmlib): rpmlib(CompressedFileNames) <= 3.0.4-1 rpmlib(FileDigests) <= 4.6.0-1 rpmlib(PartialHardlinkSets) <= 4.0.4-1 rpmlib(PayloadFilesHavePrefix) <= 4.0-1
Checking for unpackaged file(s): /usr/lib/rpm/check-files /home/user/rpmbuild/BUILDROOT/xen-4.8.5-6.fc25.x86_64
Wrote: pkgs/dom0-fc25/x86_64/xen-4.8.5-6.fc25.x86_64.rpm
Wrote: pkgs/dom0-fc25/x86_64/python2-xen-4.8.5-6.fc25.x86_64.rpm
Wrote: pkgs/dom0-fc25/x86_64/python3-xen-4.8.5-6.fc25.x86_64.rpm
Wrote: pkgs/dom0-fc25/x86_64/xen-libs-4.8.5-6.fc25.x86_64.rpm
Wrote: pkgs/dom0-fc25/x86_64/xen-runtime-4.8.5-6.fc25.x86_64.rpm
Wrote: pkgs/dom0-fc25/x86_64/xen-hypervisor-4.8.5-6.fc25.x86_64.rpm
Wrote: pkgs/dom0-fc25/x86_64/xen-doc-4.8.5-6.fc25.x86_64.rpm
Wrote: pkgs/dom0-fc25/x86_64/xen-devel-4.8.5-6.fc25.x86_64.rpm
Wrote: pkgs/dom0-fc25/x86_64/xen-licenses-4.8.5-6.fc25.x86_64.rpm
Wrote: pkgs/dom0-fc25/x86_64/xen-hvm-4.8.5-6.fc25.x86_64.rpm
Wrote: pkgs/dom0-fc25/x86_64/xen-qemu-tools-4.8.5-6.fc25.x86_64.rpm
Wrote: pkgs/dom0-fc25/x86_64/xen-qubes-vm-4.8.5-6.fc25.x86_64.rpm
Wrote: pkgs/dom0-fc25/x86_64/xen-debuginfo-4.8.5-6.fc25.x86_64.rpm
Executing(%clean): /bin/sh -e /var/tmp/rpm-tmp.NmoYu3
+ umask 022
+ cd /home/user/rpmbuild/BUILD
+ cd xen-4.8.5
+ rm -rf /home/user/rpmbuild/BUILDROOT/xen-4.8.5-6.fc25.x86_64
+ exit 0
--> Done:
      qubes-src/vmm-xen/pkgs/dom0-fc25/x86_64/python2-xen-4.8.5-6.fc25.x86_64.rpm
      qubes-src/vmm-xen/pkgs/dom0-fc25/x86_64/python3-xen-4.8.5-6.fc25.x86_64.rpm
      qubes-src/vmm-xen/pkgs/dom0-fc25/x86_64/xen-4.8.5-6.fc25.x86_64.rpm
      qubes-src/vmm-xen/pkgs/dom0-fc25/x86_64/xen-debuginfo-4.8.5-6.fc25.x86_64.rpm
      qubes-src/vmm-xen/pkgs/dom0-fc25/x86_64/xen-devel-4.8.5-6.fc25.x86_64.rpm
      qubes-src/vmm-xen/pkgs/dom0-fc25/x86_64/xen-doc-4.8.5-6.fc25.x86_64.rpm
      qubes-src/vmm-xen/pkgs/dom0-fc25/x86_64/xen-hvm-4.8.5-6.fc25.x86_64.rpm
      qubes-src/vmm-xen/pkgs/dom0-fc25/x86_64/xen-hypervisor-4.8.5-6.fc25.x86_64.rpm
      qubes-src/vmm-xen/pkgs/dom0-fc25/x86_64/xen-libs-4.8.5-6.fc25.x86_64.rpm
      qubes-src/vmm-xen/pkgs/dom0-fc25/x86_64/xen-licenses-4.8.5-6.fc25.x86_64.rpm
      qubes-src/vmm-xen/pkgs/dom0-fc25/x86_64/xen-qemu-tools-4.8.5-6.fc25.x86_64.rpm
      qubes-src/vmm-xen/pkgs/dom0-fc25/x86_64/xen-qubes-vm-4.8.5-6.fc25.x86_64.rpm
      qubes-src/vmm-xen/pkgs/dom0-fc25/x86_64/xen-runtime-4.8.5-6.fc25.x86_64.rpm
make[2]: Leaving directory '/home/user/builder'
--> build failed!
make[1]: *** [Makefile.generic:180: packages] Error 1
make[1]: Leaving directory '/home/user/builder'
make: *** [Makefile:231: vmm-xen-dom0] Error 1
```


[patch-vesa-4G.txt](https://github.com/QubesOS/qubes-issues/files/3328863/patch-vesa-4G.txt)

**Comment by marmarek** [comment]
Interesting, there is very explicit `exit 0` in the fragment you've pasted... I'd try those packages anyway. But first inspect the build log to confirm the patch was indeed applied, just to be sure.

**Comment by JarrahG** [comment]
I've solved the issue above, qubes is now compiling with your Xen patch included.

This seems to have solved part of the problem, however not all. I am now past the initial xen boot screen outlined in the first message, but the system hangs there on a blank screen. The same blank screen occurred with just about all iterations of the UEFI troubleshooting guides options. The only change is that the system will power down when the following lines are included:
```
noexitboot=1
mapbs=1
```

I have enabled DEBUG and VERBOSE flags in builder.conf and added loglvl=all to the xen options line. Is there any other way to get debug information to print?

**Comment by JarrahG** [comment]
Using the above PR, XEN now boots (though the screen moves very slowly) and the system gets to "relinquishing VGA to dom0". 

However at this point the system stops again, back to the blank screen. This one is seemingly a dom0 issue. 

Would it be possible to get xorg-x11-drv-intel from fc29 or fc30 into dom0? Should this be considered a separate issue?

**Comment by marmarek** [comment]
Add `xen=current,keep guest_loglvl=all` options to Xen, to not give the console to dom0 but to keep Xen printing logs. You can also add `console=hvc0` to Linux, to have it forward messages to Xen (and hopefully Xen will print them). It may be that Dom0 crashes early at the boot...

**Comment by JarrahG** [comment]
Looks like you were right. Dom0 kernel panics at the 269 second mark. Screenshot attached. 

I'm compiling a new ISO with kernel 4.19. Hopefully that will make a difference.

![Dom0 Kernel Panic](https://user-images.githubusercontent.com/43260698/60338171-c58be500-9994-11e9-98f0-5f613f692bc4.jpg)

I also get a number of warnings about failing to assign PCI device memory. 

![PCI no space for...](https://user-images.githubusercontent.com/43260698/60338259-03890900-9995-11e9-9c18-8f8c835fbee2.jpg)

**Comment by marmarek** [comment]
> I also get a number of warnings about failing to assign PCI device memory.

Something went wrong with memory map, 0x1000 isn't much.

**Comment by JarrahG** [comment]
That was an ACPI issue. Disabling ACPI (dom0 commandline: acpi=off) resolved it. 

Kernel panic still occurs though. 
![IMG_20190628_212550](https://user-images.githubusercontent.com/43260698/60339282-ddb13380-9997-11e9-89dc-b66d11e8c61d.jpg)

**Comment by JarrahG** [comment]
Tested on Linux 4.19. Exactly the same issue as above. Kernel panic in xen_hypercall_xen_version. 

Not sure where to go from here. Potentially the patches above are causing Xen issues?

**Comment by marmarek** [comment]
Try xen 4.8.5-7 from current-testing (should be built in about 1h). It may be about the patches, namely bumping `__XEN_LATEST_INTERFACE_VERSION__`.

**Comment by JarrahG** [comment]
Thanks for fixing that. It's my second time going into XEN code, this one has been a bit more complicated. 

I've got it compiled myself, but I'm still getting exactly the same kernel panic seen above. I'll grab the current-testing version tomorrow morning and give it a go there, just to make sure there's nothing wrong with my build system.

**Comment by JarrahG** [comment]
Tested on the newly released 4.0.2-rc1 ISO which seems to include Xen 4.8.5-7. Exactly the same kernel panic as above. Xen boots fine with the video above 4G patch, but dom0 kernel panics on boot. 

For refencene, I am using the following boot configuration:
```
[qubes-verbose]
options=console=vga efi=no-rs iommu=no-igfx loglvl=all dom0_mem=min:1024M dom0_mem=max:4096M ucode=scan vga=current,keep guest_loglvl=all
#noexitboot=1
#mapbs=1
kernel=vmlinuz inst.stage2=hd:LABEL=Qubes-20190628-x86_64 i915.alpha_support=1 nouveau.modeset=0 iommu=no-igfx console=hvc0 acpi=off
ramdisk=initrd.img
```

**Comment by JarrahG** [comment]
Closing as the issue referred to in the title has been resolved. Xen boots on this laptop, but Dom0 kernel does not. I'll open and link a separate issue for this issue.

**Comment by firstabor** [comment]
I have the same issue with the Installer R4.0.3 out of the box, when booting in UEFI only mode.

But the UEFI only mode works with the Qubes-20200113-X86-64 Version 4.1 just fine.
What was fixed in that release that is broken in the R4.0.3? maybe this helps to solve this issue.
